### PR TITLE
Fetch project and user details server side

### LINKED ISSUE #249: When project is shared on socials etc - it takes the share metadata of the project
Image NFT = image
Metadate Title = title project

Just to update this so its promoting the project vs launch
Screenshot 2023-08-29 at 10 24 04

### PR DESCRIPTION
**Pull request by thisisommore**
fix #249

### CODE REVIEW

**Comment by thisisommore** [comment]
@marcuspang here

**Comment by marcuspang** [comment]
> @marcuspang using staleTime to avoid refetching, but that is dirty workaround, any ideas what can be done there? Maybe I will experiment with staleTime 1 ms

Was there any issue with just using the project and user data from the server?

**Comment by thisisommore** [comment]
> > @marcuspang using staleTime to avoid refetching, but that is dirty workaround, any ideas what can be done there? Maybe I will experiment with staleTime 1 ms
> 
> Was there any issue with just using the project and user data from the server?

It would refetch it even in client, and thus the meta tags will only be available after loading is finished, and og tags are expected to be returned in first request.

**Comment by thisisommore** [comment]
@marcuspang 
Okay there was another reason for using initialData, the child components are also using getProjects and getUser and if I didn't set initial set then the api will be called again and this will further slow down page
getProject
image
getUser
image

**Comment by thisisommore** [comment]
Okay so slateTime: 1 is dirty workaround or else we will have to setup tanstack query, because [queryClient.setQueryData](https://tanstack.com/query/v4/docs/react/reference/QueryClient#queryclientsetquerydata) not working properly because I think because wagmi is using it internally.

File: hooks/useGetUser.ts [hunk 1]
```diff
@@ -2,8 +2,10 @@ import { CacheKey } from '@/constants/react-query';
 import { getUser } from '@/lib/backend';
 import { useQuery } from 'wagmi';
```
Comment (marcuspang) [inline hooks/useGetUser.ts:3]:
```suggestion
import { CacheKey } from '@/constants/react-query';
import { getUser } from '@/lib/backend';
import type { User } from '@/types/mongo';
import { useQuery } from 'wagmi';
```

File: hooks/useGetUser.ts [hunk 2]
```diff
@@ -2,8 +2,10 @@ import { CacheKey } from '@/constants/react-query';
...
 
-export function useGetUser(userId?: string) {
+export function useGetUser(userId?: string, initialData?: Awaited<ReturnType<typeof getUser>>) {
```
Comment (marcuspang) [inline hooks/useGetUser.ts:5]:
```suggestion
export function useGetUser(userId?: string, initialData?: User) {
```

File: pages/project/[id]/index.tsx [hunk 3]
```diff
@@ -27,21 +27,42 @@ import Link from 'next/link';
...
+
+export const getServerSideProps = (async (context) => {
+  const id = context.params?.id as string;
+  const project_data = await getProject(id!);
+  const user_data = await getUser(project_data.founder!);
+  return { props: { project: project_data, user: user_data } };
+}) satisfies GetServerSideProps<{
+  project: Project | undefined;
+  user: Awaited<ReturnType<typeof getUser>>;
+}>;
```
Comment (marcuspang) [inline pages/project/[id]/index.tsx:42]:
```suggestion
export const getServerSideProps: GetServerSideProps<{
  project?: Project;
  user?: User;
}> = async (context) => {
  const { id } = context.params ?? {};
  if (!id) {
    return { props: { project: undefined, user: undefined } };
  }
  const project = await getProject(id.toString());
  const user = await getUser(project.founder);
  return { props: { project, user } };
};
```

File: pages/project/[id]/index.tsx [hunk 4]
```diff
@@ -27,21 +27,42 @@ import Link from 'next/link';
 import { useRouter } from 'next/router';
 import { useState } from 'react';
 
+import type { InferGetServerSidePropsType, GetServerSideProps } from 'next';
+import { Project } from '@/types/mongo';
+import { getProject, getUser } from '@/lib/backend';
```
Comment (marcuspang) [inline pages/project/[id]/index.tsx:32]:
```suggestion
import { getProject, getUser } from '@/lib/backend';
import { Project, User } from '@/types/mongo';
import type { GetServerSideProps, InferGetServerSidePropsType } from 'next';
```

File: hooks/useGetProject.ts [hunk 5]
```diff
@@ -1,9 +1,12 @@
 import { CacheKey } from '@/constants/react-query';
 import { getProject } from '@/lib/backend';
+import { Project } from '@/types/mongo';
```
Comment (marcuspang) [inline hooks/useGetProject.ts:3]:
```suggestion
import type { Project } from '@/types/mongo';
```

**Review by marcuspang — APPROVED** [review]
Some minor nits, but LGTM

**Pull request merged**